### PR TITLE
Add presentationTimestamp to RTCEncodedVideoFrameMetadata

### PR DESCRIPTION
**Pull request by tonyherre**
Add a captureTimestamp to RTCEncodedVideoFrameMetadata, defined only for local device captures to match the mediaTime as defined in requestVideoFrameCallback() and also match the `timestamp` in WebCodecs VideoFrame and EncodedVideoChunk. This allows apps to match encoded frames after encoding/before decoding with the corresponding raw frames in the [mediacapture-transform](https://www.w3.org/TR/mediacapture-transform/) APIs.

Essentially an adoption of the open https://github.com/w3c/webrtc-encoded-transform/pull/137, incorporating the comments there.

***
<a href="https://pr-preview.s3.amazonaws.com/tonyherre/webrtc-encoded-transform/pull/173.html" title="Last updated on Apr 24, 2023, 1:36 PM UTC (f661a21)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/webrtc-encoded-transform/173/5c7ab84...tonyherre:f661a21.html" title="Last updated on Apr 24, 2023, 1:36 PM UTC (f661a21)">Diff</a>

### CODE REVIEW

**Comment by tonyherre** [comment]
@alvestrand Uploaded as we had discussed. Needs some editorial work to add links in the right places, but the logic is there. PTAL

File: index.bs [hunk 1]
```diff
@@ -365,6 +366,18 @@ dictionary RTCEncodedVideoFrameMetadata {
...
+        <dfn>captureTimestamp</dfn> of type <span class=
+            "idlMemberType">long long</span>
+    </dt>
+    <dd>
+        <p>
+		For video frames coming from a local device, this is the timestamp in microseconds
+		at which the frame was captured by the camera.
+		This will have the same value as the basis for VideoFrameCallbackMetadata's captureTime
+		field.
+        </p>
```
Comment (alvestrand) [inline index.bs:379]: Should mention that this is defined to match WebCodecs.

File: index.bs [hunk 2]
```diff
@@ -365,6 +366,17 @@ dictionary RTCEncodedVideoFrameMetadata {
...
     </dd>
+    <dt>
+        <dfn>presentationTimestamp</dfn> of type <span class=
+            "idlMemberType">long long</span>
+    </dt>
+    <dd>
+        <p>
+            The media presentation timestamp (PTS) in microseconds of raw frame, matching the
+	    {{VideoFrame/timestamp}} for raw frames which correspond to this frame and the
+	    {{VideoFrameCallbackMetadata/mediaTime}} given if this frame is decoded and rendererd.
```
Comment (youennf) [inline index.bs:377]: VideoFrameCallbackMetadata/mediaTime is a double in seconds, maybe we should remove the reference to this one.

**Comment by aboba** [comment]
Why is `presentationTimestamp` defined with a different type than `presentationTime` as defined in [VideoFrameCallbackMetadata](https://wicg.github.io/video-rvfc/#video-frame-metadata-callback)?

File: index.bs (hunk 2 above)
Comment (jan-ivar) [reply index.bs:377]: > VideoFrameCallbackMetadata/mediaTime is a double in seconds, maybe we should remove the reference to this one.

Do you mean milliseconds? I also found https://github.com/w3c/webcodecs/issues/122 on why WebCodecs uses microseconds for timestamps.

Comment (youennf) [reply index.bs:377]: https://wicg.github.io/video-rvfc/#dom-videoframecallbackmetadata-mediatime

Comment (jan-ivar) [reply index.bs:377]: Ah, it uses seconds to match [mediaElement.currentTime](https://html.spec.whatwg.org/multipage/media.html#dom-media-currenttime). 🤦

While I think it makes sense for a method on [HTMLVideoElement](https://html.spec.whatwg.org/multipage/media.html#htmlvideoelement) to stay consistent with its attributes, this here is arguably a lower-level API.

To make matters complicated, W3C design principle [§ 8.3. Use milliseconds for time measurement](https://w3ctag.github.io/design-principles/#milliseconds) say... that.

And in https://github.com/w3c/webcodecs/issues/122 they use microseconds, out of concern for audio AV drift.

I see the original PR was for `captureTimestamp` and then it was changed to `presentationTimestamp`... What's the use case for this value? We may need to consider this to figure out which unit to stay consistent with.

Comment (youennf) [reply index.bs:377]: Our model is VideoFrame -> encoder -> encoded chunk -> encoded transform.
Seems best to be consistent with VideoFrame here, and remove the reference to rvfc

Comment (jan-ivar) [reply index.bs:377]: I see the OP was edited and originally used microseconds. Can someone clarify what changed? Sorry if I missed it, I find the github thread here a bit hard to follow.

Normally, an issue is opened first for discussion, which tends to leave a thread that is more easy to follow, then a PR is opened later once discussion solidifies (github's PR workflow tends to hide things once resolved, which suits a review process more than a discussion imho).

Comment (tonyherre) [reply index.bs:377]: Thanks for the discussion and links!
Consistency with VideoFrame is indeed the motivating usecase here as Youenn said - so that an app is able to associate a raw frame before encoding with an encoded frame afterwards, when applying transforms on both sides. I'm happy to just remove the reference to rVFC here - would definitely simplify the text.

Apologies for editing the in-flight PR. The intention was always to have a field here which matched VideoFrame.timestamp, but the previous PR https://github.com/w3c/webrtc-encoded-transform/pull/137 that I was asked to adopt seemed to mistake which timestamp this corresponded to. Given noone but Harald had commented here yet, I thought it easier to just modify in place before kicking off the review more widely.

**Comment by tonyherre** [comment]
> Why is `presentationTimestamp` defined with a different type than `presentationTime` as defined in [VideoFrameCallbackMetadata](https://wicg.github.io/video-rvfc/#video-frame-metadata-callback)? If these are defining the same thing, can we be consistent?

The primary motivator is for being able to match with VideoFrame.timestamp, which uses a long long timestamp. I was trying to also pull in the rVFC field, per the discussion towards the end of https://github.com/w3c/webrtc-encoded-transform/pull/137, but that was probably a mistake. Per the comment thread with Jan-Ivar and Youenn, it seems to make sense to define this only in reference with the VideoFrame timestamp - I've pushed a commit removing the ref from the text at least. Does that make sense to you @aboba?

**Comment by jan-ivar** [comment]
Thanks, microseconds makes sense now. However, I remain concerned about the name.

Consistency with [VideoFrame](https://w3c.github.io/webcodecs/#dom-videoframe-timestamp) and [EncodedVideoChunk](https://w3c.github.io/webcodecs/#dom-encodedvideochunkinit-timestamp) would mean calling it _"`timestamp` ... The presentation timestamp, given in microseconds. The timestamp is copied from the [EncodedVideoChunk](https://w3c.github.io/webcodecs/#encodedvideochunk) corresponding to this VideoFrame."_

However, in [RTCEncodedVideoFrame](https://w3c.github.io/webrtc-encoded-transform/#rtcencodedvideoframe) `timestamp` is taken (used for RTP). But there's a comment claiming this _"will eventually re-use or extend the equivalent defined in WebCodecs"_ (presumably [EncodedVideoChunk](https://w3c.github.io/webcodecs/#encodedvideochunk))?

image

If that is still our intent, we would appear to be at a crossroads. Is it too late to align this?

**Comment by tonyherre** [comment]
How about us using `timestamp` here inside RTCEncodedVideoFrameMetadata for this presentation timestamp field, to match WebCodecs, and I'll create a follow-up PR to also add `rtpTimestamp` inside this Metadata and deprecating/removing the `RTCEncodedVideoFrame.timestamp` field?

This moves us into more alignment with the WebCodecs type by having a consistent `timestamp` field, gets away from the confusion of having two `timestamp`s without needing to have an immediate breaking change, and preserves the amount of information available to web apps.

**Review by henbos — APPROVED** [review]

**Review by jan-ivar — APPROVED** [review]
